src/provider/provider_coarse.c

```diff
@@ -0,0 +1,1405 @@
+    }
+
+    *provider = coarse_provider;
+
+    assert(debug_check(coarse_provider));
+
+    return UMF_RESULT_SUCCESS;
+
+block_error:
+    ravl_tree_rm(coarse_provider->upstream_alloc, coarse_provider->init_buffer);
+err_delete_ravl_all_blocks:
+    ravl_delete(coarse_provider->all_blocks);
+err_delete_ravl_free_blocks:
+    ravl_delete(coarse_provider->free_blocks);
+err_delete_ravl_upstream_alloc:
+    ravl_delete(coarse_provider->upstream_alloc);
+err_free_coarse_provider:
+    umf_ba_global_free(coarse_provider);
+    return ret;
+}
+
+static void ravl_cb_rm_upstream_alloc_node(void *data, void *arg) {
+    assert(data);
+    assert(arg);
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)arg;
+    ravl_data_t *node_data = data;
+    block_t *alloc = node_data->value;
+    assert(alloc);
+
+    if (coarse_provider->upstream_memory_provider &&
+        coarse_provider->WA_do_not_free_upstream == false) {
+        umf_result_t ret =
+            umfMemoryProviderFree(coarse_provider->upstream_memory_provider,
+                                  alloc->data, alloc->size);
+
+        // We would continue to deallocate alloc blocks even if the upstream
+        // provider doesn't return success.
+        assert(ret == UMF_RESULT_SUCCESS);
+        (void)ret;
+    }
+
+    assert(coarse_provider->alloc_size >= alloc->size);
+    coarse_provider->alloc_size -= alloc->size;
+
+    umf_ba_global_free(alloc);
+}
+
+static void ravl_cb_rm_all_blocks_node(void *data, void *arg) {
+    assert(data);
+    assert(arg);
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)arg;
+    ravl_data_t *node_data = data;
+    block_t *block = node_data->value;
+    assert(block);
+
+    if (block->used) {
+        assert(coarse_provider->used_size >= block->size);
+        coarse_provider->used_size -= block->size;
+    }
+
+    if (block->free_list_ptr) {
+        free_blocks_rm_node(coarse_provider->free_blocks, block->free_list_ptr);
+    }
+
+    umf_ba_global_free(block);
+}
+
+static void coarse_memory_provider_finalize(void *provider) {
+    if (provider == NULL) {
+        assert(0);
+        return;
+    }
+
+    coarse_memory_provider_t *coarse_provider =
+        (struct coarse_memory_provider_t *)provider;
+
+    ravl_foreach(coarse_provider->upstream_alloc,
+                 ravl_cb_rm_upstream_alloc_node, coarse_provider);
+    assert(coarse_provider->alloc_size == 0);
+
+    ravl_foreach(coarse_provider->all_blocks, ravl_cb_rm_all_blocks_node,
+                 coarse_provider);
+    assert(coarse_provider->used_size == 0);
+
+    ravl_delete(coarse_provider->upstream_alloc);
+    ravl_delete(coarse_provider->all_blocks);
+    ravl_delete(coarse_provider->free_blocks);
+
+    umf_ba_global_free(coarse_provider);
+}
+
+static umf_result_t coarse_memory_provider_alloc(void *provider, size_t size,
```

Comment:
this function is too long, please split it into multiple function

Reply:
Fixed in https://github.com/oneapi-src/unified-memory-framework/pull/715